dotCMS/src/main/java/com/dotcms/cost/RequestCostFilter.java

```diff
@@ -0,0 +1,132 @@
+package com.dotcms.cost;
+
+import com.dotcms.cost.RequestCostApi.Accounting;
+import com.dotmarketing.business.APILocator;
+import com.liferay.util.servlet.NullServletResponse;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+
+public class RequestCostFilter implements Filter {
+
+
+    private final RequestCostApi requestCostApi;
+    private final LeakyTokenBucket bucket;
+
+    public RequestCostFilter() {
+        this(APILocator.getRequestCostAPI(), new LeakyTokenBucket());
+    }
+
+    RequestCostFilter(RequestCostApi requestCostApi) {
+        this(requestCostApi, new LeakyTokenBucket());
+    }
+
+
+    RequestCostFilter(RequestCostApi requestCostApi, LeakyTokenBucket bucket) {
+        this.requestCostApi = requestCostApi;
+        this.bucket = bucket;
+
+    }
+
+
+    @Override
+    public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
+            throws IOException, ServletException {
+
+        HttpServletRequest request = (HttpServletRequest) req;
+        HttpServletResponse response = (HttpServletResponse) res;
+
+        Accounting fullAccounting = requestCostApi.resolveAccounting(request);
+
+        boolean allowed = bucket.allow();
+        response.addHeader("X-dotRateLimit-Toks/Max", bucket.getTokenCount() + "/" + bucket.maximumBucketSize);
+
+        if (!allowed) {
+            response.sendError(429);
+            return;
+        }
+
+        HttpServletResponse wrapper =
+                fullAccounting == Accounting.HTML ? new NullServletResponse(response)
+                        : new RequestCostResponseWrapper(request, response);
+        requestCostApi.addCostHeader(request, wrapper);
+        chain.doFilter(req, wrapper);
+        requestCostApi.addCostHeader(request, wrapper);
```

Comment:
Why do we add the CostHeader before and after?

Reply:
This and the ResponseWrapper were "tricks" to try to get the header to write properly once a response has been committed, but I don't think either of these actually work.